keps/sig-scheduling/20190221-even-pods-spreading.md

```diff
@@ -0,0 +1,423 @@
+
+Inside `TopologySpreadConstraint`, we need hard affinityTerms (similar with
+`PodAffinityTerm`) and soft affinityTerms (similar with
+`WeightedPodAffinityTerm`). This describes when we perform even distribution,
+which pods are considered as a group.
+
+```go
+type TopologySpreadConstraint struct {
+    // MaxSkew describes the degree of imbalance of pods spreading.
+    // It's the max difference between the number of matching pods in any two
+    // topology domains of a given topology type.
+    // Default value is 1 and 0 is not allowed.
+    MaxSkew int32
+    // TopologyKey defines where pods are placed evenly
+    TopologyKey string
+    // Similar with the same field in PodAffinity/PodAntiAffinity
+    // +optional
+    RequiredDuringSchedulingIgnoredDuringExecution []PodAffinityTerm
+    // Similar with the same field in PodAffinity/PodAntiAffinity
+    // +optional
+    PreferredDuringSchedulingIgnoredDuringExecution []WeightedPodAffinityTerm
+}
+```
+
+#### Option 2
+
+Another option is to flatten "required" and "perferred" podAffinityTerms, and
+eliminate embedded "TopologyKey":
+
+```go
+type ScheduleDecision string
+
+const (
+    // do not schedule a pod in all circumstances
+    DoNotSchedule ScheduleDecision = "DoNotSchedule"
+    // schedule a pod despite of any circumstance
+    ScheduleAnyway ScheduleDecision = "ScheduleAnyway"
+)
+
+type TopologySpreadConstraint struct {
+    // MaxSkew describes the degree of imbalance of pods spreading.
+    // It's the max difference between the number of matching pods in any two
+    // topology domains of a given topology type.
+    // For example, in a 3-zone cluster, currently pods with the same labelSelector
+    // are spread as 1/1/0:
+    // - if MaxSkew is 1, incoming pod can only be scheduled to zone3 to become 1/1/1;
+    // schedule it onto zone1(zone2) will make the ActualSkew(2) violates MaxSkew(1)
+    // - if MaxSkew is 2, incoming pod can be scheduled to any zone.
+    // Default value is 1 and 0 is not allowed.
+    MaxSkew int32
+    // TopologyKey is the key such that we consider each value as a "bucket";
+    // we try to put balanced number of pods into each bucket.
+    TopologyKey string
+    // WhenUnsatisfiable indicates how to deal with a pod if it doesn't satisfy
+    // the spreading constraint.
+    // - DoNotSchedule (default) tells the scheduler not to schedule it
+    // - ScheduleAnyway tells the scheduler to still schedule it
+    // Note: it's considered as "Unsatisfiable" only when ActualySkew on all nodes exceeds "MaxSkew".
+    WhenUnsatisfiable ScheduleDecision
+    // Label selector for pods. This's enforced by scheduler to check which pods
+    // should be recognized as a group to satisfy the spreading constraint.
+    Selector *metav1.LabelSelector
+    // namespaces specifies which namespaces the labelSelector applies to (matches against);
+    // null or empty list means "this pod's namespace"
+    // +optional
+    Namespaces []string
```

Comment:
I still don't see a user story or something explaining why you would ever want a different namespace than the pod's own namespace?

Reply:
I used to put it as a derivation from PodAffinity. I agree we can remove it and default to "this pod's namespace".

cc @bsalamat in case Bobby happen to know some histories or particular user story.